AX: Stitch adjacent text elements together in the accessibility tree, in turn requiring ATs to do less work and providing a better user experience during element navigation

twilco opened the pull request:
#### eb755a4654c3b939d8da1f74ac5aba81be427d73
<pre>
AX: Stitch adjacent text elements together in the accessibility tree, in turn requiring ATs to do less work and providing a better user experience during element navigation
<a href="https://bugs.webkit.org/show_bug.cgi?id=302894">https://bugs.webkit.org/show_bug.cgi?id=302894</a>
<a href="https://rdar.apple.com/165159758">rdar://165159758</a>

Reviewed by Joshua Hoffman.

Prior to this commit, the WebKit accessibility tree exposes text in a way that matches the DOM and render tree. This
provides a poor user experience, because the presence of spans or other elements, or even just whitespace can result
in text nodes being split, despite the visual rendering still looking contiguous. As a concrete example, this simple text
requires 8 swipes for iOS VoiceOver to traverse through:

```
&lt;span&gt;Hello&lt;/span&gt; &lt;span&gt;world&lt;/span&gt;
&lt;br/&gt;
&lt;span&gt;Foo&lt;/span&gt; &lt;span&gt;&lt;a href=&quot;#foo&quot;&gt;bar&lt;/a&gt;&lt;/span&gt; &lt;span&gt;baz&lt;/span&gt; &lt;span&gt;last&lt;/span&gt;
```

With this commit, we implement a new feature called accessibility text stitching, which does what it sounds like --
stitches adjacent text elements into one, and prevents the exposure of the stitched text in the accessibility tree.

After this commit, that same example only requires three swipes — one for the text before the link, one for the link,
and one for the text after the link.

Specfically, this commit introduces the concept of `stithedUnignoredChildren`, and uses it in place of unignored children in some
places (only where necessary, depending on the context, since computing stitched children is even more work on top of
unignored children).

Test accessibility/dynamic-text-stitching.html added. This is also effectively tested by many other layout tests which
expect string values of static text elements (our existing test suite pointed me at many bugs which I fixed).

* LayoutTests/accessibility/dynamic-text-stitching-expected.txt: Added.
* LayoutTests/accessibility/dynamic-text-stitching.html: Added.
* LayoutTests/resources/accessibility-helper.js:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::transformToFinalChildren):
(WebCore::AXCoreObject::finalChildren):
(WebCore::AXCoreObject::blockFlowAncestor const):
(WebCore::AXCoreObject::stitchStateFromGroups const):
(WebCore::AXCoreObject::crossFrameFinalChildren):
(WebCore::AXCoreObject::contents):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::isBlockFlow const):
(WebCore::AXCoreObject::hasStitchableRole const):
(WebCore::AXCoreObject::blockFlowAncestorForStitchable const):
(WebCore::AXCoreObject::StitchState::StitchState):
(WebCore::AXCoreObject::stitchState const):
(WebCore::AXCoreObject::stitchedIntoObject const):
* Source/WebCore/accessibility/AXID.h: Added.
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::AXObjectCache):
(WebCore::AXObjectCache::stitchGroupsOwnedBy):
(WebCore::AXObjectCache::onLaidOutInlineContent):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::appendChildrenToArray):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::isBlockFlow const):
(WebCore::hasEnclosingInputElement):
(WebCore::hasStitchBreakingRole):
(WebCore::hasStitchBreakingTag):
(WebCore::isStitchBreakingElement):
(WebCore::shouldStopStitchingAt):
(WebCore::AccessibilityNodeObject::stitchGroups const):
(WebCore::AccessibilityNodeObject::stitchState const):
(WebCore::AccessibilityNodeObject::stringValue const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::stitchGroups const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::stringValue const):
(WebCore::AccessibilityRenderObject::boundingBoxRect const):
* Source/WebCore/accessibility/StitchGroup.cpp: Added.
(WebCore::StitchGroup::lastNode const):
(WebCore::StitchGroup::lastObject const):
* Source/WebCore/accessibility/StitchGroup.h: Added.
(WebCore::StitchGroup::StitchGroup):
(WebCore::StitchGroup::vector const):
(WebCore::StitchGroups::StitchGroups):
(WebCore::StitchGroups::vector const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityTextMarker accessibilityObject]):
(-[WebAccessibilityObjectWrapper accessibilityElements]):
(-[WebAccessibilityObjectWrapper accessibilityElementCount]):
(-[WebAccessibilityObjectWrapper accessibilityElementAtIndex:]):
(-[WebAccessibilityObjectWrapper indexOfAccessibilityElement:]):
(-[WebAccessibilityObjectWrapper accessibilityRowRange]):
(-[WebAccessibilityObjectWrapper containsUnnaturallySegmentedChildren]):
(-[WebAccessibilityObjectWrapper accessibilityObjectForTextMarker:]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::isDefaultValue):
(WebCore::AXIsolatedObject::stitchGroupsView const):
(WebCore::AXIsolatedObject::stitchState const):
(WebCore::AXIsolatedObject::relativeFrame const):
(WebCore::AXIsolatedObject::stringValue const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::convertToPropertyFlag):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(children):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
(-[WebAccessibilityObjectWrapper accessibilityIndexOfChild:]):
(-[WebAccessibilityObjectWrapper accessibilityArrayAttributeCount:]):
(-[WebAccessibilityObjectWrapper _accessibilityChildrenFromIndex:maxCount:returnPlatformElements:]):
* Source/WebCore/layout/integration/inline/InlineIteratorBox.h:
(WebCore::InlineIterator::Box::isAtomicInlineBox const):
* Source/WebCore/layout/integration/inline/InlineIteratorBoxLegacyPath.h:
(WebCore::InlineIterator::BoxLegacyPath::isAtomicInlineBox const):
* Source/WebCore/layout/integration/inline/InlineIteratorBoxModernPath.h:
(WebCore::InlineIterator::BoxModernPath::isAtomicInlineBox const):
* Source/WebCore/page/DeprecatedGlobalSettings.h:
(WebCore::DeprecatedGlobalSettings::setAccessibilityTextStitchingEnabled):
(WebCore::DeprecatedGlobalSettings::accessibilityTextStitchingEnabled):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutInlineContent):

Canonical link: <a href="https://commits.webkit.org/303574@main">https://commits.webkit.org/303574@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e8ba76e424ecf56835e7615014743d81e4dc026e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132870 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5371 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43959 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140407 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84899 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134740 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5739 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/5234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101600 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68910 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135816 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119045 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82395 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83636 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124940 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112921 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37164 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143057 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/131378 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5040 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37748 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109973 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5122 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/5234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110152 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27922 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3859 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115311 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58571 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5094 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33665 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/164345 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4933 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68546 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42663 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5184 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5052 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on LayoutTests/accessibility/dynamic-text-stitching.html:
```diff
@@ -0,0 +1,38 @@
+if (window.accessibilityController) {
+    window.jsTestIsAsync = true;
+
+    var text = accessibilityController.rootElement.childAtIndex(0).childAtIndex(0);
```
I dislike debugging chains of childAtIndex when a test fails because it's hard to know where the breakage happened. How about:

var body = accessibilityController.rootElement.childAtIndex(0);   <-- or whatever this element actually corresponds to
var staticText = body.childAtIndex(0);  // container is presentational, so this is the first unignored child

minorninth commented on LayoutTests/resources/accessibility-helper.js:
```diff
@@ -70,6 +70,10 @@ function dumpAXSearchTraversal(axElement, options = { }) {
     return output;
 }
 
+function platformTextValue(axElement) {
```
How about platformStaticTextValue so it's less ambiguous. You could also optionally assert the role here

minorninth commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -360,6 +360,54 @@ AXCoreObject* AXCoreObject::firstUnignoredChild()
 
 #endif // ENABLE(INCLUDE_IGNORED_IN_CORE_AX_TREE)
 
+static AXCoreObject::AccessibilityChildrenVector transformToFinalChildren(AXCoreObject::AccessibilityChildrenVector&& children)
```
How about childrenAfterStitching or filterStitchedNodesFromChildren

minorninth commented on Source/WebCore/accessibility/AXCoreObject.h:
```diff
@@ -1011,6 +1011,44 @@ class AXCoreObject : public RefCountedAndCanMakeWeakPtr<AXCoreObject> {
+        }
+    };
+    virtual StitchState stitchState() const { return { }; }
+    StitchState stitchStateFromGroups(const StitchGroups*) const;
```
Should this be private? It feels like an implementation detail, not something that should be called from another object

twilco replied:
The subclasses do need it, so I moved to protected!

minorninth commented on Source/WebCore/accessibility/AXCoreObject.h:
```diff
@@ -1011,6 +1011,44 @@ class AXCoreObject : public RefCountedAndCanMakeWeakPtr<AXCoreObject> {
+    };
+    virtual StitchState stitchState() const { return { }; }
+    StitchState stitchStateFromGroups(const StitchGroups*) const;
+    std::optional<AXID> stitchedIntoObject() const
```
Should this be stitchedIntoID()

minorninth commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -906,6 +920,8 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
     VisibleSelection m_lastSelection;
 #endif
+
+    WeakHashMap<RenderObject, StitchGroups, SingleThreadWeakPtrImpl> m_stitchGroups;
```
Note main thread only?

minorninth replied:
(Unless this is already part of a section marked as main thread only)

twilco replied:
I'm pretty sure that there are zero methods or member variables that are safe to use off the main-thread, so I think this may be implicit. But maybe I'm forgetting something obvious 🤔

minorninth commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -3924,6 +3927,194 @@ String AccessibilityNodeObject::text() const
+
+static bool isStitchBreakingElement(Element& element)
+{
+    return is<HTMLTableCellElement>(element)
```
What about something with display: table-cell, display: inline-block, or really anything other than display: inline?

twilco replied:
I think those would affect the layout of inline content, and thus we get the right behavior "for free".

minorninth commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -3924,6 +3927,194 @@ String AccessibilityNodeObject::text() const
+        return true;
+    }
+
+    if (hasEnclosingInputElement(renderer.node())) {
```
What about textarea and contenteditable? I feel like it'd be safer to not stitch contenteditable, but I'm not sure. We should test with editing a complex email. That would definitely be a good stress test if things were being constantly stitched and unstitched as you type.

twilco replied:
Filed issue for us to follow-up on this question.

minorninth commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -3924,6 +3927,194 @@ String AccessibilityNodeObject::text() const
+    return node && is<HTMLInputElement>(node->shadowHost());
+}
+
+static bool hasStitchBreakingRole(Element& element)
```
AccessibilityNodeObject is so large already, what would you think of having StitchUtils.h/cpp for most of these static helper functions?

If you're worried about performance, could we put them all in something like StitchUtils-Inline.h?

hoffmanjoshua replied:
+1 to this idea!

minorninth commented on Source/WebCore/accessibility/StitchGroup.h:
```diff
@@ -0,0 +1,72 @@
+        : m_group(WTFMove(group))
+    { }
+
+    const FixedVector<AXID>& vector() const { return m_group; }
```
How about something more semantic, like objectIDs?

minorninth commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -379,6 +427,11 @@ AXCoreObject::AccessibilityChildrenVector AXCoreObject::crossFrameUnignoredChild
     return result;
 }
 
+AXCoreObject::AccessibilityChildrenVector AXCoreObject::crossFrameFinalChildren()
```
I'm not convinced we need this. A cross frame child should be something like a AXWebArea, there's never any case where it should have stitching.

I'd prefer removing it and instead add an assert somewhere, like maybe when calling crossFrameUnignoredChildren we could add an assert that none of the children are stitched into anything

minorninth reviewed (commented):
The implementation looks pretty clean and readable, and the complexity seems well contained! Overall seems pretty close to ready to merge.

Instead of finalChildren I think we should be more specific about it including stitching. It's already hard enough for someone encountering the code the first time to know which of many children methods to call.

Maybe in AXCoreObject.h there should be a comment explaining the "levels" of children.

aria-owns could mess things up, maybe opt out of stitching if aria-owns is set anywhere

hoffmanjoshua commented on Source/WebCore/accessibility/StitchGroup.h:
```diff
@@ -25,22 +25,48 @@
+class AXObjectCache;
+class AccessibilityObject;
+
+class StitchGroup : public CanMakeCheckedPtr<StitchGroup> {
```
To follow the pattern of our other classes, do you think we can name this `AXStitchGroup`, or `AccessibilityStitchGroup`?

hoffmanjoshua commented on Source/WebCore/accessibility/AXCoreObject.h:
```diff
@@ -1012,6 +1011,44 @@ class AXCoreObject : public RefCountedAndCanMakeWeakPtr<AXCoreObject> {
+        // stitched into it.
+        const std::optional<AXID> stitchedIntoID { std::nullopt };
+        // The stitch group that |A| belongs to, if any.
+        const CheckedPtr<const StitchGroup> group { nullptr };
```
I wonder if it would be better to name this "ownerGroup", or someting similar? Reading group above wasn't the most intuitive until I read the comment here. What do you think?

twilco replied:
Not opposed to changing it, but I don't understand the idea behind "owner group". This struct represents the stitch state for an object — what is the object stitched into (`stitchedIntoID`), and what stitch `group` (hence the current name) does the object belong to. Maybe you can explain more about your owner group idea or come up with a name that you think would be more intuitive? IMO this reads well at the call sites without being too wordy — `stitchState.group`.

hoffmanjoshua replied:
Yeah, `ownerGroup` doesn't seem like an appropriate name either. I think the use of "belongs" is what trips me up. there is some object that will return the contents of the group (`stitchedIntoID`), which is what I am thinking about when I think "ownership", but that might not be necessary to convey here. I think it's fine as-is—thanks for the explanation

twilco replied:
> there is some object that will return the contents of the group (stitchedIntoID), which is what I am thinking about when I think "ownership"

Ahh OK yes, your suggestion makes more sense to me now! Owner is such a loaded term with `aria-owns` unfortunately. Yeah I think let's leave it for now but let me know if you come up with another idea that you like better 👍

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -3937,8 +4128,25 @@ String AccessibilityNodeObject::stringValue() const
+        CheckedPtr cache = axObjectCache();
+
+        RefPtr endNode = stitchState.group && cache ? stitchState.group->lastNode(*cache) : nullptr;
+        if (!endNode || endNode == node)
```
Is it possible for this block to get hit for the `endNode == node` case? In other words, is a sttich group with just a single element possible?

twilco replied:
> is a stitch group with just a single element possible?

It's not. But what theoretically could happen is `lastNode` could fail every `objectForID` in the group up until `this`, meaning `node` would equal `endNode`, so this tries to guard against that.

hoffmanjoshua reviewed (approved):
All LGTM, r+

twilco committed:
AX: Stitch adjacent text elements together in the accessibility tree, in turn requiring ATs to do less work and providing a better user experience during element navigation
https://bugs.webkit.org/show_bug.cgi?id=302894
rdar://165159758

Reviewed by Joshua Hoffman.

Prior to this commit, the WebKit accessibility tree exposes text in a way that matches the DOM and render tree. This
provides a poor user experience, because the presence of spans or other elements, or even just whitespace can result
in text nodes being split, despite the visual rendering still looking contiguous. As a concrete example, this simple text
requires 8 swipes for iOS VoiceOver to traverse through:

```
<span>Hello</span> <span>world</span>
<br/>
<span>Foo</span> <span><a href="#foo">bar</a></span> <span>baz</span> <span>last</span>
```

With this commit, we implement a new feature called accessibility text stitching, which does what it sounds like --
stitches adjacent text elements into one, and prevents the exposure of the stitched text in the accessibility tree.

After this commit, that same example only requires three swipes — one for the text before the link, one for the link,
and one for the text after the link.

Specfically, this commit introduces the concept of `stithedUnignoredChildren`, and uses it in place of unignored children in some
places (only where necessary, depending on the context, since computing stitched children is even more work on top of
unignored children).

Test accessibility/dynamic-text-stitching.html added. This is also effectively tested by many other layout tests which
expect string values of static text elements (our existing test suite pointed me at many bugs which I fixed).

* LayoutTests/accessibility/dynamic-text-stitching-expected.txt: Added.
* LayoutTests/accessibility/dynamic-text-stitching.html: Added.
* LayoutTests/resources/accessibility-helper.js:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::transformToFinalChildren):
(WebCore::AXCoreObject::finalChildren):
(WebCore::AXCoreObject::blockFlowAncestor const):
(WebCore::AXCoreObject::stitchStateFromGroups const):
(WebCore::AXCoreObject::crossFrameFinalChildren):
(WebCore::AXCoreObject::contents):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::isBlockFlow const):
(WebCore::AXCoreObject::hasStitchableRole const):
(WebCore::AXCoreObject::blockFlowAncestorForStitchable const):
(WebCore::AXCoreObject::StitchState::StitchState):
(WebCore::AXCoreObject::stitchState const):
(WebCore::AXCoreObject::stitchedIntoObject const):
* Source/WebCore/accessibility/AXID.h: Added.
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::AXObjectCache):
(WebCore::AXObjectCache::stitchGroupsOwnedBy):
(WebCore::AXObjectCache::onLaidOutInlineContent):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::appendChildrenToArray):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::isBlockFlow const):
(WebCore::hasEnclosingInputElement):
(WebCore::hasStitchBreakingRole):
(WebCore::hasStitchBreakingTag):
(WebCore::isStitchBreakingElement):
(WebCore::shouldStopStitchingAt):
(WebCore::AccessibilityNodeObject::stitchGroups const):
(WebCore::AccessibilityNodeObject::stitchState const):
(WebCore::AccessibilityNodeObject::stringValue const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::stitchGroups const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::stringValue const):
(WebCore::AccessibilityRenderObject::boundingBoxRect const):
* Source/WebCore/accessibility/StitchGroup.cpp: Added.
(WebCore::StitchGroup::lastNode const):
(WebCore::StitchGroup::lastObject const):
* Source/WebCore/accessibility/StitchGroup.h: Added.
(WebCore::StitchGroup::StitchGroup):
(WebCore::StitchGroup::vector const):
(WebCore::StitchGroups::StitchGroups):
(WebCore::StitchGroups::vector const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityTextMarker accessibilityObject]):
(-[WebAccessibilityObjectWrapper accessibilityElements]):
(-[WebAccessibilityObjectWrapper accessibilityElementCount]):
(-[WebAccessibilityObjectWrapper accessibilityElementAtIndex:]):
(-[WebAccessibilityObjectWrapper indexOfAccessibilityElement:]):
(-[WebAccessibilityObjectWrapper accessibilityRowRange]):
(-[WebAccessibilityObjectWrapper containsUnnaturallySegmentedChildren]):
(-[WebAccessibilityObjectWrapper accessibilityObjectForTextMarker:]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::isDefaultValue):
(WebCore::AXIsolatedObject::stitchGroupsView const):
(WebCore::AXIsolatedObject::stitchState const):
(WebCore::AXIsolatedObject::relativeFrame const):
(WebCore::AXIsolatedObject::stringValue const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::convertToPropertyFlag):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(children):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
(-[WebAccessibilityObjectWrapper accessibilityIndexOfChild:]):
(-[WebAccessibilityObjectWrapper accessibilityArrayAttributeCount:]):
(-[WebAccessibilityObjectWrapper _accessibilityChildrenFromIndex:maxCount:returnPlatformElements:]):
* Source/WebCore/layout/integration/inline/InlineIteratorBox.h:
(WebCore::InlineIterator::Box::isAtomicInlineBox const):
* Source/WebCore/layout/integration/inline/InlineIteratorBoxLegacyPath.h:
(WebCore::InlineIterator::BoxLegacyPath::isAtomicInlineBox const):
* Source/WebCore/layout/integration/inline/InlineIteratorBoxModernPath.h:
(WebCore::InlineIterator::BoxModernPath::isAtomicInlineBox const):
* Source/WebCore/page/DeprecatedGlobalSettings.h:
(WebCore::DeprecatedGlobalSettings::setAccessibilityTextStitchingEnabled):
(WebCore::DeprecatedGlobalSettings::accessibilityTextStitchingEnabled):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutInlineContent):

Canonical link: https://commits.webkit.org/303574@main
